pkg/cosign/verify.go

```diff
@@ -620,20 +620,71 @@ func verifySignatures(ctx context.Context, sigs oci.Signatures, h v1.Hash, co *C
 func verifyInternal(ctx context.Context, sig oci.Signature, h v1.Hash,
 	verifyFn signatureVerificationFn, co *CheckOpts) (
 	bundleVerified bool, err error) {
+	var acceptableRFC3161Time, acceptableRekorBundleTime *time.Time // Timestamps for the signature we accept, or nil if not applicable.
+
+	if co.TSACerts != nil {
+		acceptableRFC3161Timestamp, err := VerifyRFC3161Timestamp(ctx, sig, co.TSACerts)
+		if err != nil {
+			return false, fmt.Errorf("unable to verify RFC3161 timestamp bundle: %w", err)
+		}
+		if acceptableRFC3161Timestamp != nil {
+			bundleVerified = true
+			acceptableRFC3161Time = &acceptableRFC3161Timestamp.Time
+		}
+	}
+
+	if !co.SkipTlogVerify {
+		bundleVerified, err = VerifyBundle(ctx, sig, co, co.RekorClient)
+		if err != nil {
+			return false, fmt.Errorf("error verifying bundle: %w", err)
+		}
+
+		if bundleVerified {
+			// Update with the verified bundle's integrated time.
+			t, err := getBundleIntegratedTime(sig)
+			if err != nil {
+				return false, fmt.Errorf("error getting bundle integrated time: %w", err)
+			}
+			acceptableRekorBundleTime = &t
+		} else {
+			// If the --offline flag was specified, fail here. bundleVerified returns false with
+			// no error when there was no bundle provided.
+			// TODO: You can be offline when you are verifying with a public key. This shouldn't
+			// error out, but maybe should be a log message.
+			if co.Offline {
+				return false, fmt.Errorf("offline verification failed")
+			}
+
+			if co.RekorClient != nil {
```

Comment:
moving my comment over here, but I think this might need to fail, since this will implicitly allow tlog verification to be skipped if there is no RekorClient. I think it would be an error if this were nil at this stage

Reply:
+1

Reply:
Combined with https://github.com/sigstore/cosign/pull/2504#discussion_r1038508344 , if this function succeeds, `bundleVerified == !co.SkipTlogVerify`.

I think that’s actually good; the `bundleVerified` return value should be removed entirely. “verify signature” should give a yes/no answer based on a submitted policy, not a menu for the poor external caller to make a choice (which different callers making different choices).

Reply:
> Combined with https://github.com/sigstore/cosign/pull/2504#discussion_r1038508344 , if this function succeeds, bundleVerified == !co.SkipTlogVerify.

That's a nice invariant!